### PR TITLE
[SPARK-21070][PYSPARK] Attempt to update cloudpickle again

### PR DESCRIPTION
**Pull request by holdenk**
## What changes were proposed in this pull request?

Based on https://github.com/apache/spark/pull/18282 by @rgbkrk this PR attempts to update to the current released cloudpickle and minimize the difference between Spark cloudpickle and "stock" cloud pickle with the goal of eventually using the stock cloud pickle.

Some notable changes:
* Import submodules accessed by pickled functions (cloudpipe/cloudpickle#80)
* Support recursive functions inside closures (cloudpipe/cloudpickle#89, cloudpipe/cloudpickle#90)
* Fix ResourceWarnings and DeprecationWarnings (cloudpipe/cloudpickle#88)
* Assume modules with __file__ attribute are not dynamic (cloudpipe/cloudpickle#85)
* Make cloudpickle Python 3.6 compatible (cloudpipe/cloudpickle#72)
* Allow pickling of builtin methods (cloudpipe/cloudpickle#57)
* Add ability to pickle dynamically created modules (cloudpipe/cloudpickle#52)
* Support method descriptor (cloudpipe/cloudpickle#46)
* No more pickling of closed files, was broken on Python 3 (cloudpipe/cloudpickle#32)
* ** Remove non-standard __transient__check (cloudpipe/cloudpickle#110)** -- while we don't use this internally, and have no tests or documentation for its use, downstream code may use __transient__, although it has never been part of the API, if we merge this we should include a note about this in the release notes.
* Support for pickling loggers (yay!) (cloudpipe/cloudpickle#96)
* BUG: Fix crash when pickling dynamic class cycles. (cloudpipe/cloudpickle#102)

## How was this patch tested?

Existing PySpark unit tests + the unit tests from the cloudpickle project on their own.

### CODE REVIEW

**Comment by holdenk** [comment]
cc @rgbkrk since this is based on your original PR. + @HyukjinKwon

**Comment by rgbkrk** [comment]
You are my hero, thank you.

**Comment by HyukjinKwon** [comment]
I guess this should be fine in general if it simply ports the changes in cloudpickle. Will double check next week (to me, it takes a while to double check).

@rgbkrk, would you mind double checking this one and see if it looks good to you?

**Comment by rgbkrk** [comment]
Yeah this is looking great, it also brings in the latest from cloudpickle that wasn't in my original patch.

File: python/pyspark/cloudpickle.py [hunk 1]
```diff
@@ -397,42 +625,7 @@ def save_global(self, obj, name=None, pack=struct.pack):
...
-            if getattr(obj, '_is_namedtuple_', False):
-                self.save_reduce(_load_namedtuple, (obj.__name__, obj._fields))
-                return
-
-            self.save(_load_class)
-            self.save_reduce(typ, (obj.__name__, obj.__bases__, {"__doc__": obj.__doc__}), obj=obj)
-            d.pop('__doc__', None)
-            # handle property and staticmethod
-            dd = {}
-            for k, v in d.items():
```
Comment (HyukjinKwon) [inline python/pyspark/cloudpickle.py]: I assume we are fine per the tests added in https://github.com/apache/spark/commit/e044705b4402f86d0557ecd146f3565388c7eeb4.

**Review by HyukjinKwon — APPROVED** [review]
@holdenk, LGTM with few questions.

What I checked was few custom changes after https://github.com/apache/spark/commit/04e44b37cc04f62fbf9e08c7076349e0a4d12ea8 assuming these were started from both https://github.com/cloudpipe/cloudpickle/commit/c4f885116126b2ac49deae7a31d4941d006f319f and https://github.com/cloudpipe/cloudpickle/commit/7aebb7ed42258a9392c2ada9b4bb390d566630cc

I checked https://github.com/apache/spark/commit/e044705b4402f86d0557ecd146f3565388c7eeb4, https://github.com/apache/spark/commit/04e44b37cc04f62fbf9e08c7076349e0a4d12ea8, https://github.com/apache/spark/commit/e044705b4402f86d0557ecd146f3565388c7eeb4, https://github.com/apache/spark/commit/55204181004c105c7a3e8c31a099b37e48bfd953, https://github.com/apache/spark/commit/ee913e6e2d58dfac20f3f06ff306081bd0e48066, https://github.com/apache/spark/commit/d48935400ca47275f677b527c636976af09332c8, https://github.com/apache/spark/commit/dbfc7aa4d0d5457bc92e1e66d065c6088d476843, https://github.com/apache/spark/commit/20e6280626fe243b170a2e7c5e018c67f3dac1db and https://github.com/apache/spark/commit/6297697f975960a3006c4e58b4964d9ac40eeaf5 which I assume we added after that.

File: python/pyspark/cloudpickle.py (hunk 1 above)
Comment (holdenk) [inline python/pyspark/cloudpickle.py]: Lets double check this part with @davies.

File: python/pyspark/cloudpickle.py [hunk 2]
```diff
@@ -220,12 +322,7 @@ def save_function(self, obj, name=None):
 
         if name is None:
             name = obj.__name__
-        try:
-            # whichmodule() could fail, see
-            # https://bitbucket.org/gutworth/six/issues/63/importing-six-breaks-pickling
-            modname = pickle.whichmodule(obj, name)
-        except Exception:
-            modname = None
+        modname = pickle.whichmodule(obj, name)
```
Comment (holdenk) [inline python/pyspark/cloudpickle.py]: It's a good question, the underlying issue was marked resolved in 2014 and from looking at the commit it seems like it should actually be resolved. That being said its true some people might be on a system installed with an old verison of six so perhaps we should keep this workaround.

**Comment by HyukjinKwon** [comment]
BTW, I also checked it passes tests with Python 3.6 in my local.

**Comment by holdenk** [comment]
If we can reach agreement on this I'll see about trying to get our local workarounds upstreamed into cloudpickle.

**Comment by rgbkrk** [comment]
Workarounds welcomed on cloudpickle. 😄

**Review by BryanCutler — APPROVED** [review]
Thanks for doing this @holdenk , LGTM!

**Comment by rgbkrk** [comment]
Just a note that we just shipped the fixes from @HyukjinKwon within cloudpickle (as v0.4.0), so we're at least roughly in sync now. 😄

File: python/pyspark/cloudpickle.py (hunk 1 above)
Comment (holdenk) [inline python/pyspark/cloudpickle.py]: Gentle re-ping to @davies - do you have an opinion on this?

**Comment by holdenk** [comment]
huzzah! I'm in the middle of getting some code working for a talk tomorrow so I'll circle back on this on Friday. If @davies has any opinions though it would be great to hear them.

**Comment by ueshin** [comment]
Thank you for cc-ing me.
I looked over this and it looks good to me except for a comment https://github.com/apache/spark/pull/18734#discussion_r130763298.
I was not exactly sure if we could remove these.

**Comment by rgbkrk** [comment]
> I looked over this and it looks good to me except for a comment #18734 (comment). I was not exactly sure if we could remove these.

cloudpickle now has this workaround, so it can stay.

File: python/pyspark/cloudpickle.py [hunk 3]
```diff
@@ -241,11 +338,32 @@ def save_function(self, obj, name=None):
...
+                else:
+                    raise pickle.PicklingError("Can't pickle %r" % obj)
+            return Pickler.save_reduce(self, obj=obj, *rv)
+
         # if func is lambda, def'ed at prompt, is in main, or is nested, then
         # we'll pickle the actual function object rather than simply saving a
         # reference (as is done in default pickler), via save_function_tuple.
-        if islambda(obj) or obj.__code__.co_filename == '<stdin>' or themodule is None:
-            #print("save global", islambda(obj), obj.__code__.co_filename, modname, themodule)
+        if (islambda(obj)
```
Comment (HyukjinKwon) [inline python/pyspark/cloudpickle.py]: Just as a side note, it looks this PR includes https://github.com/cloudpipe/cloudpickle/pull/51 too (SPARK-21753).

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Let's go with this. I am quite sure it is worth for a go. We can almost sync it and can deduplucate review costs and etc.. Will go merging this one in few days with another double check if there is no objection here and it is not mergrd by anyone.

**Review by rgbkrk — APPROVED** [review]

**Comment by holdenk** [comment]
Sounds good @HyukjinKwon :) I think we should consider if we want to upstream the named tuple workaround as well.

**Comment by holdenk** [comment]
But for now yes I think we are in a good position. If no one objects of course :)

**Comment by HyukjinKwon** [comment]
Yea, it looks so. Named tuple one reminds me of the workaround we have for named tuple to make picklable - https://github.com/apache/spark/blob/d03aebbe6508ba441dc87f9546f27aeb27553d77/python/pyspark/serializers.py#L395-L446

Maybe, we could take a look and see if we could get rid of this or port this. Anyway, let me take a final look.

**Comment by HyukjinKwon** [comment]
I am merging this because:

cloudpickle looks initially ported from https://github.com/cloudpipe/cloudpickle/commit/7aebb7ed42258a9392c2ada9b4bb390d566630cc and https://github.com/cloudpipe/cloudpickle/commit/c4f885116126b2ac49deae7a31d4941d006f319f (-> https://github.com/apache/spark/commit/04e44b37cc04f62fbf9e08c7076349e0a4d12ea8), where I see both are identical.

After https://github.com/apache/spark/commit/04e44b37cc04f62fbf9e08c7076349e0a4d12ea8, we have diff - https://github.com/apache/spark/commit/e044705b4402f86d0557ecd146f3565388c7eeb4, https://github.com/apache/spark/commit/55204181004c105c7a3e8c31a099b37e48bfd953, https://github.com/apache/spark/commit/ee913e6e2d58dfac20f3f06ff306081bd0e48066, https://github.com/apache/spark/commit/d48935400ca47275f677b527c636976af09332c8, https://github.com/apache/spark/commit/dbfc7aa4d0d5457bc92e1e66d065c6088d476843, https://github.com/apache/spark/commit/20e6280626fe243b170a2e7c5e018c67f3dac1db and https://github.com/apache/spark/commit/6297697f975960a3006c4e58b4964d9ac40eeaf5

**[SPARK-9116] [SQL] [PYSPARK] support Python only UDT in __main__**, https://github.com/apache/spark/commit/e044705b4402f86d0557ecd146f3565388c7eeb4: I think this part is only what we are worried of. It looks supporting `classmethod`, `staticmethod` and `property`. We have a test:

https://github.com/apache/spark/blob/96608310501a43fa4ab9f2697f202d655dba98c5/python/pyspark/sql/tests.py#L141-L173

https://github.com/apache/spark/blob/96608310501a43fa4ab9f2697f202d655dba98c5/python/pyspark/sql/tests.py#L898-L927

**[SPARK-10542] [PYSPARK] fix serialize namedtuple**, https://github.com/apache/spark/commit/55204181004c105c7a3e8c31a099b37e48bfd953: We keep the changes:

https://github.com/holdenk/spark/blob/f986c2591a9a0b6962862c5cdfc33a7d65be7eda/python/pyspark/cloudpickle.py#L1090-L1095

https://github.com/holdenk/spark/blob/f986c2591a9a0b6962862c5cdfc33a7d65be7eda/python/pyspark/cloudpickle.py#L433-L436

and the related test pass:

https://github.com/apache/spark/blob/77cc0d67d5a7ea526f8efd37b2590923953cb8e0/python/pyspark/tests.py#L211-L219

**[SPARK-13697] [PYSPARK] Fix the missing module name of TransformFunctionSerializer.loads**, https://github.com/apache/spark/commit/ee913e6e2d58dfac20f3f06ff306081bd0e48066: We keep this change:

https://github.com/holdenk/spark/blob/f986c2591a9a0b6962862c5cdfc33a7d65be7eda/python/pyspark/cloudpickle.py#L528

https://github.com/holdenk/spark/blob/f986c2591a9a0b6962862c5cdfc33a7d65be7eda/python/pyspark/cloudpickle.py#L1022-L1029

and the related test pass:

https://github.com/apache/spark/blob/77cc0d67d5a7ea526f8efd37b2590923953cb8e0/python/pyspark/tests.py#L233-L237

We should probably port this one to `cloudpipe/cloudpickle`.


**[SPARK-16077] [PYSPARK] catch the exception from pickle.whichmodule()**, https://github.com/apache/spark/commit/d48935400ca47275f677b527c636976af09332c8: We keep this change:

https://github.com/holdenk/spark/blob/f986c2591a9a0b6962862c5cdfc33a7d65be7eda/python/pyspark/cloudpickle.py#L325-L330

https://github.com/holdenk/spark/blob/f986c2591a9a0b6962862c5cdfc33a7d65be7eda/python/pyspark/cloudpickle.py#L620-L625

This patch even should be safer as I and @rgbkrk verified this with some tests:

https://github.com/cloudpipe/cloudpickle/pull/112


**[SPARK-17472] [PYSPARK] Better error message for serialization failures of large objects in Python**, https://github.com/apache/spark/commit/dbfc7aa4d0d5457bc92e1e66d065c6088d476843: We keep this change:

https://github.com/holdenk/spark/blob/f986c2591a9a0b6962862c5cdfc33a7d65be7eda/python/pyspark/cloudpickle.py#L240-L249

Probably, we should port this change into `cloudpipe/cloudpickle`.

**[SPARK-19019] [PYTHON] Fix hijacked `collections.namedtuple` and port cloudpickle changes for PySpark to work with Python 3.6.0**, https://github.com/apache/spark/commit/20e6280626fe243b170a2e7c5e018c67f3dac1db

This change was ported from `cloudpipe/cloudpickle`. I tested our PySpark tests pass with Python 3.6.0 in my local manually - https://github.com/apache/spark/pull/18734#issuecomment-319558550

**[SPARK-19505][PYTHON] AttributeError on Exception.message in Python3**, https://github.com/apache/spark/commit/6297697f975960a3006c4e58b4964d9ac40eeaf5: We keep this change:

https://github.com/holdenk/spark/blob/f986c2591a9a0b6962862c5cdfc33a7d65be7eda/python/pyspark/cloudpickle.py#L240-L249

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
LGTM

I am going to credit this to @rgbkrk per (http://spark.apache.org/contributing.html)

> In case several people contributed, prefer to assign to the more ‘junior’, non-committer contributor

I just double checked if the tests passes with Python 3.6.0, and if I could run pi example with pypy manually (SPARK-21753), with the current status.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by ushine** [comment]
Can you guys please exclude me from the loop?

-
Best,
Daniel Biesiada

**Comment by HyukjinKwon** [comment]
Oh, it was me cc'ing you mistakenly. Sorry for the noise.

**Pull request closed without merging**